numaproj/numaplane
# Avoid data loss when pipeline, ISBService, or Numaflow Controller is updated

Fixes #9 
Fixes #137 

### Modifications

The change is as described in [these slides](https://docs.google.com/presentation/d/1XtP-k5ny1RFEWUFgkROFgp4WcPdc_9CWgTDGL0A2NE0/edit#slide=id.p).

Basically, pausing of pipelines occurs in these cases:

- Pipeline needs to update
- ISBService needs to update
- Numaflow Controller needs to update

In order for the ISBService to be able to update, the following occurs:

1. ISBServiceRO reconciler indicates in memory its need for pausing and then enqueues all Pipelines using it to be processed by PipelineRO reconciliation queue
2. PipelineRO reconciliation worker sees ISBService's request for pausing and pauses pipeline
3. ISBServiceRO reconciler waits for all pipelines to be paused
4. Once they are paused, it proceeds to update
5. Once it has verified that the ISBService plus its underlying StatefulSet is completely updated, it sets its own need for pausing to true and enqueues the Pipelines
6. PipelineRO reconciliation worker sees that nobody has a need for pausing anymore and resumes pipeline

The Numaflow Controller RO reconciler follows a similar process.

### Verification

Only added a few unit tests, where the functionality was complicated. Wasn't terribly useful to add to our Gingko unit tests as explained [here](https://github.com/numaproj/numaplane/pull/125/files#r1686737460), as this feature introduces more interdependencies between the different rollouts and I'm thinking it may make more sense to move some of this unit test logic into the e2e test.

I did do a lot of manual testing, however. I tested both with the feature flag on and off.

**Testing with feature flag off:**
Performed basic CRUD (creation, update, deletion) and verified that the Status phase got set and Conditions.

**Testing with feature flag on:**
Procedure:

1. Create NumaflowControllerRollout, ISBServiceRollout, PipelineRollout
8. Edit each one and observe that the Pipeline gets paused prior to update occurring, and then resumes
9. Verified Status looks correct, including that `ChildResourcesDeployed.ObservedGeneration` matches the generation in which the change occurred

### Remaining Items

- add more testing (I think it mostly should occur in e2e but we can discuss the unit testing thing)
- do a manual test in which I can confirm there is no data loss when doing all of these updates
- address all of the TODOs I added (one of them is that I'd like to if possible have a single struct for "existing Object" and for "new Object" which includes everything in it - the GenericObject form, the Unstructured form, the Pipeline/ISBService spec form and pass that in everywhere, so that we can avoid doing too many conversions back and forth between these)
